File: Leetcode100/leetcode100_python/TreeSum.py
from typing import List

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        n = len(nums)
        result = []
        for i in range(n):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            target = 0 - nums[i]
            l = i + 1
            r = n - 1
            while l < r:
                if nums[l] + nums[r] < target:
                    l += 1
                elif nums[l] + nums[r] > target:
                    r -= 1
                else:
                    result.append([nums[i], nums[l], nums[r]])
                    l += 1
                    r -= 1
                    while l < r and nums[l] == nums[l - 1]:
                        l += 1
        return result


if __name__ == "__main__":
    nums = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.threeSum(nums))

File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/three_sum/TreeSum.java
package main.java.com.jiang.leetcode100.three_sum;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

/**
 * 超时了
 */
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> line = new ArrayList<>();
        qsort(nums, 0, nums.length - 1);
        treeSumHelper(nums, 0, 0, 0, line, result);
        return result;
    }

    public void qsort(int[] nums, int begin, int end) {
        if(begin >= end) return;
        int v = nums[begin];
        int i = begin, j = end;
        while(i < j) {
            while(i < j && nums[j] >= v) j--;
            nums[i] = nums[j];
            while(i < j && nums[i] <= v) i++;
            nums[j] = nums[i];
        }
        nums[i] = v;
        qsort(nums, begin, i - 1);
        qsort(nums, i + 1, end);
    }

    public void treeSumHelper(int[] nums, int index, int deepth, int sum, 
        List<Integer> line, List<List<Integer>> result) {
        if(deepth == 3) {
            if(sum == 0) {
                List<Integer> t = new ArrayList<>(line);
                result.add(t);
            }
            return;
        }

        for(int i = index; i < nums.length; i++) {
            if(i > index && nums[i] == nums[i - 1]) continue;
            line.add(nums[i]);
            treeSumHelper(nums, i + 1, deepth + 1, sum + nums[i], line, result);
            line.remove(line.indexOf(nums[i]));
        }

    }
}

public class TreeSum {
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String str = scanner.nextLine();
        scanner.close();

        String[] nums_str = str.split(" ");
        int[] nums  =new int[nums_str.length];
        for(int i = 0; i < nums_str.length; i++) {
            nums[i] = Integer.parseInt(nums_str[i]);
        }

        List<List<Integer>> result = new Solution().threeSum(nums);
        for(int i = 0; i < result.size(); i++) {
            for(int j = 0; j < result.get(i).size(); j++) {
                System.out.print(result.get(i).get(j) + " ");
            }
            System.out.println("\n");
        }
    }

}